### PR TITLE
[RayJob] Support ActiveDeadlineSeconds

### PR DESCRIPTION
**Pull request by kevin85421**
## Why are these changes needed?
* Initially, `StartTime` refers to the time when the Ray job starts running on the RayCluster. This PR redefines `StartTime` to the time when `JobDeploymentStatus` transitioned from `New` to `Initializing`.
* `ActiveDeadlineSeconds`: Similar to `ActiveDeadlineSeconds` in K8s Job https://kubernetes.io/docs/concepts/workloads/controllers/job/.
* We don't check `ActiveDeadlineSeconds` if the RayJob is in `New`, `Complete`, `Suspending`, and `Suspended`.
* We only check `ActiveDeadlineSeconds` if the RayJob is in the `Initializing` or `Running` state, which may occupy some computing resources.
  * If the RayJob passes `ActiveDeadlineSeconds`, KubeRay will transition the status to `Complete`.

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [x] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: ray-operator/controllers/ray/rayjob_controller.go [hunk 1]
```diff
@@ -175,12 +180,19 @@ func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request)
 		if shouldUpdate := r.updateStatusToSuspendingIfNeeded(ctx, rayJobInstance); shouldUpdate {
 			break
 		}
+		if pastActiveDeadline(rayJobInstance) {
+			r.Log.Info("The RayJob has passed the activeDeadlineSeconds. Transition the status to `Complete`.", "StartTime", rayJobInstance.Status.StartTime, "ActiveDeadlineSeconds", *rayJobInstance.Spec.ActiveDeadlineSeconds)
+			rayJobInstance.Status.JobDeploymentStatus = rayv1.JobDeploymentStatusComplete
+			break
+		}
 
-		// TODO (kevin85421): For light-weight mode, calculate the number of failed retries and transition
```
Comment (kevin85421) [inline ray-operator/controllers/ray/rayjob_controller.go:179]: We may consider implementing backoffLimit if ActiveDeadlineSeconds proves to be insufficient.

**Comment by kevin85421** [comment]
cc @andrewsykim would you mind reviewing this PR?

File: ray-operator/apis/ray/v1/rayjob_types.go [hunk 2]
```diff
@@ -114,7 +115,6 @@ type RayJobStatus struct {
 	RayClusterStatus RayClusterStatus `json:"rayClusterStatus,omitempty"`
 	// observedGeneration is the most recent generation observed for this RayJob. It corresponds to the
 	// RayJob's generation, which is updated on mutation by the API Server.
-	// +optional
```
Comment (andrewsykim) [inline ray-operator/apis/ray/v1/rayjob_types.go:117]: What's the reason for this change?

File: ray-operator/controllers/ray/rayjob_controller.go [hunk 3]
```diff
@@ -142,6 +142,11 @@ func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request)
 		if shouldUpdate := r.updateStatusToSuspendingIfNeeded(ctx, rayJobInstance); shouldUpdate {
 			break
 		}
+		if pastActiveDeadline(rayJobInstance) {
+			r.Log.Info("The RayJob has passed the activeDeadlineSeconds. Transition the status to `Complete`.", "StartTime", rayJobInstance.Status.StartTime, "ActiveDeadlineSeconds", *rayJobInstance.Spec.ActiveDeadlineSeconds)
+			rayJobInstance.Status.JobDeploymentStatus = rayv1.JobDeploymentStatusComplete
```
Comment (andrewsykim) [inline ray-operator/controllers/ray/rayjob_controller.go:147]: Do you think exceeding deadline seconds warrants a new status `DeadlineExceeded`?

Comment (andrewsykim) [reply ray-operator/controllers/ray/rayjob_controller.go:147]: If not -- if user sees status `Complete`, are the ray-operator logs the only way to know if a job successfully completed or timed out?

File: ray-operator/test/e2e/rayjob_test.go [hunk 4]
```diff
@@ -212,4 +211,24 @@ env_vars:
 		test.Consistently(RayJob(test, rayJob.Namespace, rayJob.Name), 5*time.Second).
 			Should(WithTransform(RayJobDeploymentStatus, Equal(rayv1.JobDeploymentStatusNew)))
 	})
+
+	test.T().Run("RayJob has passed ActiveDeadlineSeconds", func(t *testing.T) {
```
Comment (andrewsykim) [inline ray-operator/test/e2e/rayjob_test.go:215]: Also add unit test if possible?

File: ray-operator/apis/ray/v1/rayjob_types.go (hunk 2 above)
Comment (kevin85421) [reply ray-operator/apis/ray/v1/rayjob_types.go:117]: I just found that it is useless. There is no change in the CRD YAML after I remove the `+optional`.

File: ray-operator/controllers/ray/rayjob_controller.go (hunk 3 above)
Comment (kevin85421) [reply ray-operator/controllers/ray/rayjob_controller.go:147]: I will improve the observability of `Complete` before KubeRay v1.1.0 release. Last week, I chatted with the [Flyte](https://github.com/flyteorg/flyte) community, and they require sufficient observability to distinguish between "Succeeded" and "Failed" statuses, including the reasons for each (e.g. backoffLimit, app-level bug, DeadlineExceeded).

**Review by andrewsykim — APPROVED** [review]

File: ray-operator/test/e2e/rayjob_test.go (hunk 4 above)
Comment (kevin85421) [reply ray-operator/test/e2e/rayjob_test.go:215]: Update https://github.com/ray-project/kuberay/pull/1933/commits/b41820da4c0c8834b3d12944a7d0053175f25886

File: docs/reference/api.md [hunk 5]
```diff
@@ -140,6 +140,7 @@ _Appears in:_
 | `jobId` _string_ | If jobId is not set, a new jobId will be auto-generated. |
 | `shutdownAfterJobFinishes` _boolean_ | ShutdownAfterJobFinishes will determine whether to delete the ray cluster once rayJob succeed or failed. |
 | `ttlSecondsAfterFinished` _integer_ | TTLSecondsAfterFinished is the TTL to clean up RayCluster. It's only working when ShutdownAfterJobFinishes set to true. |
+| `activeDeadlineSeconds` _integer_ | ActiveDeadlineSeconds is the duration in seconds that the RayJob may be active before KubeRay actively tries to terminate the RayJob; value must be positive integer. |
```
Comment (jjyao) [inline docs/reference/api.md:143]: where is the check that value must be postive?

Comment (kevin85421) [reply docs/reference/api.md:143]: Update https://github.com/ray-project/kuberay/pull/1933/commits/b9862d1ba5258e968b789f68def6c0c1342e712c

**Review by jjyao — APPROVED** [review]

File: ray-operator/controllers/ray/rayjob_controller.go (hunk 3 above)
Comment (kevin85421) [reply ray-operator/controllers/ray/rayjob_controller.go:147]: @Yicheng-Lu-llll will take this issue.
